concurrency/MemberReceiver.java

```diff
@@ -0,0 +1,2351 @@
+    if (recorder != null) {
+        return recorder.getRecordPath();
+    }
+
+    return null;
+    }
+
+    private void recordPacket(byte[] data, int length) {
+    if (cp.getFromRecordingFile() == null) {
+        return;
+    }
+
+    if (recordRtp == false) {
+        return;
+    }
+
+    synchronized (recordingLock) {
+        if (recorder == null) {
+            return;
+        }
+
+        try {
+            recorder.writePacket(data, 0, length);
+            } catch (IOException e) {
+                Logger.println("Unable to record data " + e.getMessage());
+            cp.setFromRecordingFile(null);
+                recorder = null;
+            }
+    }
+    }
+
+    private void recordAudio(int[] data, int length) {
+        if (cp.getFromRecordingFile() == null) {
+            return;
+        }
+
+    if (recordRtp == true) {
+        return;
+    }
+
+    synchronized (recordingLock) {
+        if (recorder == null) {
+            return;
+        }
+
+            try {
+                recorder.write(data, 0, length);
+            } catch (IOException e) {
+                Logger.println("Unable to record data " + e.getMessage());
+            cp.setFromRecordingFile(null);
+                recorder = null;
+            }
+    }
+    }
+
+    public boolean doNotRecord() {
+    return cp.doNotRecord();
+    }
+
+    public void setDoNotRecord(boolean doNotRecord) {
+    cp.setDoNotRecord(doNotRecord);
+
+    Logger.println("Call " + cp + " doNotRecord is " + doNotRecord);
+    }
+
+    public void setRecordFromMember(boolean enabled, String recordingFile,
+        String recordingType) throws IOException {
+
+    if (doNotRecord()) {
+        if (enabled) {
+        Logger.println("Call " + cp + " doesn't allow recording.");
+        enabled = false;
+        }
+    }
+
+        if (recorder != null) {
+        recorder.done();
+        recorder = null;
+        }
+
+    synchronized (recordingLock) {
+        if (enabled == false) {
+            cp.setFromRecordingFile(null);
+        return;
+        }
+
+            if (recordingType == null) {
+                recordingType = "Au";
+            }
+
+        recordRtp = false;
+
+        if (recordingType.equalsIgnoreCase("Rtp")) {
+                recordRtp = true;
+            }
+
+            synchronized (recordingLock) {
+        MediaInfo m;
+
+                try {
+            if (recordingType.equalsIgnoreCase("Rtp")) {
+                        m = SdpManager.findMediaInfo(
+                            myMediaInfo.getEncoding(),
+                myMediaInfo.getSampleRate(),
+                            myMediaInfo.getChannels());
+            } else {
+            m = SdpManager.findMediaInfo(
+                RtpPacket.PCM_ENCODING,
+                conferenceManager.getMediaInfo().getSampleRate(),
+                conferenceManager.getMediaInfo().getChannels());
+            }
+                } catch (ParseException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)